src/otx/algo/callbacks/adaptive_train_scheduling.py

```diff
@@ -0,0 +1,80 @@
+
+from __future__ import annotations
+
+import logging as log
+import math
+from typing import TYPE_CHECKING
+
+from lightning import Callback
+
+if TYPE_CHECKING:
+    from lightning import LightningModule, Trainer
+
+
+class AdaptiveTrainScheduling(Callback):
+    """Adaptive Training Scheduling Hook.
+
+    Depending on the size of iteration per epoch, adaptively update the validation interval and related values.
+
+    Args:
+        max_interval: Maximum value of validation interval.
+            Defaults to 5.
+        decay: Parameter to control the interval. This value is set by manual manner.
+            Defaults to -0.025.
+    """
+
+    def __init__(self, max_interval: int = 5, decay: float = -0.025):
+        self.max_interval = max_interval
+        self.decay = decay
+        self._saved_check_val_every_n_epoch: int | None = None
+        self._saved_log_every_n_steps: int | None = None
+
+    def on_train_start(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """Execute this function at starting the train stage."""
+        max_interval = min(self.max_interval, trainer.max_epochs)
+        iter_per_epoch = len(trainer.train_dataloader)
+
+        adaptive_check_val_every_n_epoch = self._get_adaptive_interval(
+            iter_per_epoch=iter_per_epoch,
+            max_interval=max_interval,
+        )
+
+        if adaptive_check_val_every_n_epoch != trainer.check_val_every_n_epoch:
+            msg = (
+                "You are using AdaptiveTrainScheduling hook. "
+                "This hook will temporarily update Trainer.check_val_every_n_epoch adaptively: "
+                f"{trainer.check_val_every_n_epoch} => {adaptive_check_val_every_n_epoch}"
+            )
+            log.warning(msg)
+
+            self._saved_check_val_every_n_epoch = trainer.check_val_every_n_epoch
+            trainer.check_val_every_n_epoch = adaptive_check_val_every_n_epoch
+
+        if iter_per_epoch < trainer.log_every_n_steps:
+            msg = (
+                "Trainer.log_every_n_steps is higher than the number of iterations in a training epoch. "
+                "To ensure logging at the last batch, temporarily update Trainer.log_every_n_steps: "
+                f"{trainer.log_every_n_steps} => {iter_per_epoch}"
+            )
+            log.warning(msg)
+
+            self._saved_log_every_n_steps = trainer.log_every_n_steps
+            trainer.log_every_n_steps = iter_per_epoch
+
+    def on_train_end(self, trainer: Trainer, pl_module: LightningModule) -> None:
+        """Execute this function at terminating the train stage."""
+        if self._saved_check_val_every_n_epoch:
+            trainer.check_val_every_n_epoch = self._saved_check_val_every_n_epoch
+            self._saved_check_val_every_n_epoch = None
+
+        if self._saved_log_every_n_steps:
+            trainer.log_every_n_steps = self._saved_log_every_n_steps
+            self._saved_log_every_n_steps = None
```

Comment:
Do we need to keep original every_n_***?

Reply:
would be good to be as close to the Lightning terminology so the Lightning users would pick it up easily? Also this callback seems to be a modular one, maybe some users could even use it standalone outside otx?

Reply:
> Do we need to keep original every_n_***?

We want to preserve the internal states of `pl.Trainer` immutable as possible.

Reply:
> > Do we need to keep original every_n_***?
> 
> We want to preserve the internal states of `pl.Trainer` immutable as possible.

Then, could every_n_epoch be different between output config from experiment and config during experiment?

Reply:
Using this callback itself means that some Trainer configurations can be adaptively reconfigured at the runtime. That's why we invoke warnings here. However, we don't want to change the given configuration values affected by this callback itself.

Reply:
> Sorry but what is the exactly one which is far away from the Lightning terminology?

I was replying to the following:
> Do we need to keep original every_n_***?